### ISSUE TITLE
REST api - 'kind' property not always exists on entities

### ISSUE THREAD

**Issue by abonas** [kubernetes/kubernetes#3030]
'kind' property exists only on the outer entity in REST api hence providing not aligned behavior for REST api consumers.
Example: when querying for all services (/api/v1beta1/services) there's an attribute 
"kind": "ServiceList" on the collection, but in the "items" section with the actual services - none of the services has the "kind" attribute.
But when querying for a specific service, for example:
http://localhost:8080/api/v1beta1/services/kubernetes-ro
There is a "kind": "Service" on the entity.

I don't see the necessity to have the "kind" property in the first place since the user knows which resource was called by the api specific url, but even if it has a reason and is by design, the entities attributes should be identical whether retrieved as a single entity or as a collection. Otherwise the client should be adding/deleting attributes when retrieved as a collection by their collection type - which is doable, but definitely not elegant.
In other words, either "kind" should be present in each entity when retrieved in bulk AND when retrieved as a single entity, or not present at all for all cases.

**Comment by lavalamp** [comment]
Here's how this should be done.

The current rule is "kind & version are set on disk and blank in memory".

So we will change the rule to "kind & version are clear in unversioned structs, but set in versioned structs". Then we'll change the conversion routines that copy TypeMeta objects. They can look at the conversion.Scope to see what exactly they're copying to/from, and they can set the kind & apiVersion fields accordingly. (conversion.Scope should already support this, I'm pretty sure I added it recently.) 

Finally, we can remove all the code in conversion that is currently responsible for setting these values before serialization and clearing them afterwards.

This should have the net effect of making serializations faster. We'll also possibly have to update tests; we could expand the api.Semantic's equality functions if that makes sense.

**Comment by smarterclayton** [comment]
We'd want to nuke TypeMeta from internal types, correct?

----- Original Message -----

**Comment by lavalamp** [comment]
Yes, it's not needed on internal types.

**Comment by liggitt** [comment]
Not having the kind on each item in the list makes it difficult to write a UI to generically describe objects (the `kind` attr is required input for something like https://github.com/kubernetes-ui/object-describer/blob/master/object-describer.js#L42-L53)

**Comment by liggitt** [comment]
@jwforres fyi

**Comment by bgrant0607** [comment]
PR to set the kind and apiversion would be welcome. Sort of related: #3000, #6439

**Comment by bgrant0607** [comment]
cc @caesarxuchao, since it's related to the other apiVersion/kind issues

**Comment by davidxia** [comment]
Hey, are there any updates on this? I'd love to see this fixed as it's convenient to have the Kind and APIVersion for each resource. Is https://github.com/kubernetes/kubernetes/pull/63972 still the PR that'll fix this? Let me know how I can help move this along.

**Comment by hjkatz** [comment]
For those of us still looking for a concise way to add TypeMeta information to runtime objects, here you go!

```
// addTypeInformationToObject adds TypeMeta information to a runtime.Object based upon the loaded scheme.Scheme
// inspired by: https://github.com/kubernetes/cli-runtime/blob/v0.19.2/pkg/printers/typesetter.go#L41
func addTypeInformationToObject(obj runtime.Object) error {
    gvks, _, err := scheme.Scheme.ObjectKinds(obj)
    if err != nil {
        return fmt.Errorf("missing apiVersion or kind and cannot assign it; %w", err)
    }

    for _, gvk := range gvks {
        if len(gvk.Kind) == 0 {
            continue
        }
        if len(gvk.Version) == 0 || gvk.Version == runtime.APIVersionInternal {
            continue
        }
        obj.GetObjectKind().SetGroupVersionKind(gvk)
        break
    }

    return nil
}
```

**Comment by xiaolin8** [comment]
> 

Sorry, I don't really get it. Where should I put this method?

**Comment by liggitt** [comment]
I hate to reopen, but I still think this is a legitimate issue that continues to impact callers. The references to this issue continue to pile up.

**Comment by thockin** [comment]
AFAICT, it IS populated though?  Am I missing something?

```
$ k get svc -o yaml | grep -A4 ^-
- apiVersion: v1
  kind: Service
  metadata:
    annotations:
      kubectl.kubernetes.io/last-applied-configuration: |
--
- apiVersion: v1
  kind: Service
  metadata:
    creationTimestamp: "2022-08-03T23:55:41Z"
    labels:
--
- apiVersion: v1
  kind: Service
  metadata:
    creationTimestamp: "2021-10-30T23:06:34Z"
    labels:
```

**Comment by liggitt** [comment]
Yeah, kubectl is messing with the output... look at the raw response from the server. `ServiceList` top level object with `items` field containing objects without a `kind` set. When that gets populated into an informer cache, initial objects lack a `kind`, then watch events for individual updates have a `kind`, and there's super weird inconsistencies about objects in the cache having apiVersion/kind set.

**Comment by thockin** [comment]
ACK.  My bad, then.

**Comment by liggitt** [comment]
/remove-sig cli

this is an api machinery issue with the conversion/server response

**Comment by liggitt** [comment]
the comment about kubectl was just that when it was writing yaml, it was serializing objects in a way that masked the issue with what the server was returning... there's no kubectl issue here I'm aware of

**Comment by leilajal** [comment]
thanks!
/assign @leilajal @Jefftree

**Comment by Jefftree** [comment]
Summarizing this issue from the comments above, the list API omits apiVersion and kind from the returned items. We should populate those fields for external types. 

It looks like there have been attempts at fixing this: 

- https://github.com/kubernetes/kubernetes/pull/80618
- https://github.com/kubernetes/kubernetes/pull/63972

haven't really looked too deep into the PRs, but the main thing I see is that the performance impact needs to be verified and we need consensus on the approach.

**Comment by thockin** [comment]
This issue still exists.  Is it a problem we actually intend to solve?

**Comment by HiranmoyChowdhury** [comment]
made a PR related to this. If you have time please see this:
https://github.com/kubernetes/kubernetes/pull/127361
